### ISSUE TITLE
LegacyVSTSPowerShellHost.exe hanging at end of custom task execution

### ISSUE THREAD

**Issue by sheldonhull** [Microsoft/vsts-agent#1286]
## Agent version and platform
Version of your agent? 2.1.24.0, also tried on earlier 2.117.2
OS of the machine running the agent? Windows

## VSTS type and version
On-Prem TFS Update 2

## What's not working?
LegacyVSTSPowerShellHost.exe is hanging at the end of a custom task. I've added explicit commands when the script execution was finished to return, as well as trying exit. Both of the output commands show it reaches these steps. However, it hangs. 

When I tried reviewing activity I noticed LegacyVSTSPowerShellHost.exe seemed to have some opened handles waiting for unnamed files/resources. 

I started terminating all handles and couldn't get it to proceed. However, once I manually terminated LegacyVSTSPowerShellHost.exe  the job proceeded to close out without issue, and then continued. Something is causing this LegacyVSTSPowerShellHost.exe  to hang. I also noted on the "hung up version" no arguments showing up, making me think it might have opened it maybe in interactive mode or something? I saw the LegacyVSTSPowerShellHost.exe  execute the next task right after, and it showed command line arguments. I'm considering that maybe it was somehow being triggered to open an interactive type session and waiting on input or acting like a console? Not sure how to validate this though.

## Agent and Worker's diag log
I reviewed debug logs, and show all activity up to end, then it just stops waiting, so there was not log activity from the worker that I saw in debug trace info at that time.

I'll update with more detail as I find it, but wanted to get this started in case there is a simple answer/test to validate how LegacyVSTSPowerShellHost.exe  is behaving and if somehow it's triggering as another "console" type window that's hidden and waiting for further input, causing task to hang.

**Comment by sheldonhull** [comment]
I know it's probably not meant to load directly, but I did try executing 
`$Executable = "C:\agent_agent\externals\vstshost\LegacyVSTSPowerShellHost.exe" ` and found that it couldn't load `Microsoft.VisualStudio.Services.WebApi.JsonUtility`. I did a search on my system and couldn't find this dll, so not sure if related, but that is one dll I couldn't find. Doing a complete system search as well to verify in case this is useful to any support on this issue

**Comment by sheldonhull** [comment]
Still running into this after tasks finished. Wondering if there is some conflict with dbatools dlls being loaded that this isn't handling the continue on script completion due to some open reference to the module library.

**Comment by bryanmacfarlane** [comment]
Custom tasks should be written using the task lib:  https://github.com/Microsoft/vsts-task-lib

This executable was only left for existing old versions of in the box tasks released with 2015.  Custom tasks should switch to the new task lib which is out of proc and much more robust.

**Comment by sheldonhull** [comment]
Thank you, this was a big help to point me in the right direction. I've found trying to track down details on differences for powershell execution hard to find. 

I had to switch over to Get-VstsInput . All my param(var) bindings were failing at that point. Is there any article or help on all the differences to I can better prepare? 

I also couldn't figure out if I could just install the latest vsts to the module path and autoload or if I really have to include this module over and over in each task. 

Thank you!

**Comment by ericsciple** [comment]
The module must be bundled with the task.

**Comment by bryanmacfarlane** [comment]
@sheldonhull - all good?  ok to close or do you need anything else?

**Comment by sheldonhull** [comment]
@bryanmacfarlane  thank you for the followup. I've been swamped. 
I still was a bit confused on the handling of parameters, and couldn't find clarity on that. See question above. 

I migrated over to Powershell3 handler, but am finding the parameter handling confusing as [cmdletbinding()]param() wasn't working, and had to use get-vstsinput instead.

**Comment by bryanmacfarlane** [comment]
@sheldonhull - please ask this as an issue in the vsts-task-lib repo.  Closing this one for now.
